File: app/modules/team/schemas.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import Field

from app.modules.shared.schemas import DashboardModel

TEAM_STATUS_PATTERN = r"^(active|suspended)$"
TEAM_WINDOW_PATTERN = r"^(day|week|month)$"


class TeamUsageWindowResponse(DashboardModel):
    cost_usd: float = 0.0
    tokens: int = 0


class TeamUsageResponse(DashboardModel):
    day: TeamUsageWindowResponse
    week: TeamUsageWindowResponse
    month: TeamUsageWindowResponse


class TeamMemberKeyResponse(DashboardModel):
    id: str
    name: str
    key_prefix: str
    is_active: bool
    last_used_at: datetime | None = None


class TeamMemberResponse(DashboardModel):
    id: str
    name: str
    email: str | None = None
    status: str
    cost_cap_day_usd: float | None = None
    cost_cap_week_usd: float | None = None
    cost_cap_month_usd: float | None = None
    token_cap_day: int | None = None
    token_cap_week: int | None = None
    token_cap_month: int | None = None
    allowed_models: list[str] | None = None
    notes: str | None = None
    created_at: datetime
    updated_at: datetime
    usage: TeamUsageResponse
    gate: str
    keys: list[TeamMemberKeyResponse] = Field(default_factory=list)


class TeamMemberCreateRequest(DashboardModel):
    name: str = Field(min_length=1, max_length=128)
    email: str | None = Field(default=None, max_length=320)
    status: str | None = Field(default=None, pattern=TEAM_STATUS_PATTERN)
    cost_cap_day_usd: float | None = Field(default=None, gt=0)
    cost_cap_week_usd: float | None = Field(default=None, gt=0)
    cost_cap_month_usd: float | None = Field(default=None, gt=0)
    token_cap_day: int | None = Field(default=None, gt=0)
    token_cap_week: int | None = Field(default=None, gt=0)
    token_cap_month: int | None = Field(default=None, gt=0)
    allowed_models: list[str] | None = None
    notes: str | None = Field(default=None, max_length=4000)


class TeamMemberUpdateRequest(DashboardModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    email: str | None = Field(default=None, max_length=320)
    status: str | None = Field(default=None, pattern=TEAM_STATUS_PATTERN)
    cost_cap_day_usd: float | None = Field(default=None, gt=0)
    cost_cap_week_usd: float | None = Field(default=None, gt=0)
    cost_cap_month_usd: float | None = Field(default=None, gt=0)
    token_cap_day: int | None = Field(default=None, gt=0)
    token_cap_week: int | None = Field(default=None, gt=0)
    token_cap_month: int | None = Field(default=None, gt=0)
    allowed_models: list[str] | None = None
    notes: str | None = Field(default=None, max_length=4000)


class TeamMemberKeyCreateRequest(DashboardModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    expires_at: datetime | None = None


class TeamUsageModelResponse(DashboardModel):
    model: str
    cost_usd: float = 0.0
    tokens: int = 0
    requests: int = 0


class TeamUsageDayResponse(DashboardModel):
    day: date
    cost_usd: float = 0.0
    tokens: int = 0


class TeamMemberUsageResponse(DashboardModel):
    member_id: str
    window: str
    window_start: datetime
    window_end: datetime
    totals: TeamUsageWindowResponse
    models: list[TeamUsageModelResponse] = Field(default_factory=list)
    series: list[TeamUsageDayResponse] = Field(default_factory=list)


class TeamOnboardingSnippets(DashboardModel):
    windows_powershell: str
    macos_zsh: str


class TeamOnboardingResponse(DashboardModel):
    base_url: str
    snippets: TeamOnboardingSnippets

File: frontend/src/features/team/schemas.ts
import { z } from "zod";

import { ApiKeyCreateResponseSchema } from "@/features/api-keys/schemas";

export const TEAM_STATUSES = ["active", "suspended"] as const;
export type TeamMemberStatus = (typeof TEAM_STATUSES)[number];

export const TEAM_GATES = ["ok", "near_cap", "over_cap", "suspended"] as const;
export type TeamGate = (typeof TEAM_GATES)[number];

export const TEAM_WINDOWS = ["day", "week", "month"] as const;
export type TeamWindow = (typeof TEAM_WINDOWS)[number];

export const TeamUsageWindowSchema = z.object({
  costUsd: z.number().default(0),
  tokens: z.number().int().default(0),
});

export const TeamUsageSchema = z.object({
  day: TeamUsageWindowSchema,
  week: TeamUsageWindowSchema,
  month: TeamUsageWindowSchema,
});

export const TeamMemberKeySchema = z.object({
  id: z.string(),
  name: z.string(),
  keyPrefix: z.string(),
  isActive: z.boolean(),
  lastUsedAt: z.string().nullable().default(null),
});

export const TeamMemberSchema = z.object({
  id: z.string(),
  name: z.string(),
  email: z.string().nullable().default(null),
  status: z.enum(TEAM_STATUSES),
  costCapDayUsd: z.number().nullable().default(null),
  costCapWeekUsd: z.number().nullable().default(null),
  costCapMonthUsd: z.number().nullable().default(null),
  tokenCapDay: z.number().nullable().default(null),
  tokenCapWeek: z.number().nullable().default(null),
  tokenCapMonth: z.number().nullable().default(null),
  allowedModels: z.array(z.string()).nullable().default(null),
  notes: z.string().nullable().default(null),
  createdAt: z.string(),
  updatedAt: z.string(),
  usage: TeamUsageSchema,
  gate: z.enum(TEAM_GATES),
  keys: z.array(TeamMemberKeySchema).default([]),
});

export const TeamMemberListSchema = z.array(TeamMemberSchema);

const capsShape = {
  costCapDayUsd: z.number().positive().nullable().optional(),
  costCapWeekUsd: z.number().positive().nullable().optional(),
  costCapMonthUsd: z.number().positive().nullable().optional(),
  tokenCapDay: z.number().int().positive().nullable().optional(),
  tokenCapWeek: z.number().int().positive().nullable().optional(),
  tokenCapMonth: z.number().int().positive().nullable().optional(),
};

export const TeamMemberCreateRequestSchema = z.object({
  name: z.string().min(1).max(128),
  email: z.string().max(320).nullable().optional(),
  status: z.enum(TEAM_STATUSES).optional(),
  allowedModels: z.array(z.string()).nullable().optional(),
  notes: z.string().max(4000).nullable().optional(),
  ...capsShape,
});

export const TeamMemberUpdateRequestSchema = z.object({
  name: z.string().min(1).max(128).optional(),
  email: z.string().max(320).nullable().optional(),
  status: z.enum(TEAM_STATUSES).optional(),
  allowedModels: z.array(z.string()).nullable().optional(),
  notes: z.string().max(4000).nullable().optional(),
  ...capsShape,
});

export const TeamMemberKeyCreateRequestSchema = z.object({
  name: z.string().min(1).max(128).optional(),
  expiresAt: z.string().nullable().optional(),
});

export const TeamUsageModelSchema = z.object({
  model: z.string(),
  costUsd: z.number().default(0),
  tokens: z.number().int().default(0),
  requests: z.number().int().default(0),
});

export const TeamUsageDaySchema = z.object({
  day: z.string(),
  costUsd: z.number().default(0),
  tokens: z.number().int().default(0),
});

export const TeamMemberUsageSchema = z.object({
  memberId: z.string(),
  window: z.enum(TEAM_WINDOWS),
  windowStart: z.string(),
  windowEnd: z.string(),
  totals: TeamUsageWindowSchema,
  models: z.array(TeamUsageModelSchema).default([]),
  series: z.array(TeamUsageDaySchema).default([]),
});

export const TeamOnboardingSchema = z.object({
  baseUrl: z.string(),
  snippets: z.object({
    windowsPowershell: z.string(),
    macosZsh: z.string(),
  }),
});

export const TeamMemberKeyCreateResponseSchema = ApiKeyCreateResponseSchema;

export type TeamMember = z.infer<typeof TeamMemberSchema>;
export type TeamMemberKey = z.infer<typeof TeamMemberKeySchema>;
export type TeamUsageWindow = z.infer<typeof TeamUsageWindowSchema>;
export type TeamMemberCreateRequest = z.infer<typeof TeamMemberCreateRequestSchema>;
export type TeamMemberUpdateRequest = z.infer<typeof TeamMemberUpdateRequestSchema>;
export type TeamMemberKeyCreateRequest = z.infer<typeof TeamMemberKeyCreateRequestSchema>;
export type TeamMemberUsage = z.infer<typeof TeamMemberUsageSchema>;
export type TeamOnboarding = z.infer<typeof TeamOnboardingSchema>;
